downloader/downloader.go

```diff
@@ -0,0 +1,179 @@
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/kkdai/youtube"
+	"github.com/vbauerster/mpb/v5"
+	"github.com/vbauerster/mpb/v5/decor"
+)
+
+// Downloader offers high level functions to download videos into files
+type Downloader struct {
+	Client    youtube.Client
+	OutputDir string // optional directory to store the files
+}
+
+func (dl *Downloader) getOutputFile(v *youtube.Video, stream *youtube.Stream, outputFile string) (string, error) {
+
+	if outputFile == "" {
+		outputFile = SanitizeFilename(v.Title)
+		outputFile += pickIdealFileExtension(stream.MimeType)
+	}
+
+	if dl.OutputDir != "" {
+		if err := os.MkdirAll(dl.OutputDir, 0755); err != nil {
+			return "", err
+		}
+		outputFile = filepath.Join(dl.OutputDir, outputFile)
+	}
+
+	return outputFile, nil
+}
+
+//Download : Starting download video by arguments.
+func (dl *Downloader) Download(ctx context.Context, outputFile string, v *youtube.Video, stream *youtube.Stream) error {
```

Comment:
I recommend to change the signature to  (ctx, video, stream, outputFile), same in other download methods

Reply:
Oh, you're right! done